### PR TITLE
DrivAerML: 16k train/eval surface points (vs 50k baseline)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The 50k surface point default may be over-sampling the surface. 16k points reduces computation, allows more case revisits per epoch, and may improve generalization. Directive from human team: test 16k/32k/64k surface point sweeps.

## Instructions

Run champion config with 16k train AND eval surface points:

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 16000 --drivaerml-eval-surface-points 16000 \
  --max-train-batches 394 \
  --wandb_group dm-surface-points-sweep
```

No `--max-eval-batches`. Report full eval test results.

## Reporting Requirements
- Best val `surface_rel_l2_pct` + epoch
- Test `surface_rel_l2_pct` at best val checkpoint (full eval)
- W&B run ID

## Baseline

**DrivAerML baselines:**
- val best: `3.997%` @ ep467 (PR #2898, W&B: `bht6h42t`, 4L/512d, AdamW lr=5e-4, T_max=30)
- test best: `6.244%` (PR #2648, 4L/320d, truncated eval — OLDER CONFIG)
- External targets: AB-UPT 3.82% | Transolver-3: 3.71%

```bash
# DrivAerML champion (val 3.997% @ ep467)
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### DrivAerML — 16k surface points (train + eval)

| Metric | Baseline (50k pts, ep467) | This run (16k pts, ep17) |
|--------|--------------------------|--------------------------|
| Val `surface_rel_l2_pct` | **3.997%** | 24.039% |
| Test `surface_rel_l2_pct` | 6.244% | 23.310% |
| Best epoch | 467 | 17 |

**External targets:** AB-UPT 3.82% | Transolver-3 3.71%

### Key finding: timeout-limited, severely underfit

The run only completed **17 epochs in the 360-minute timeout**. The model was still improving rapidly (63% → 24% over 17 epochs, loss still declining) but nowhere near converged. This makes the results **not directly comparable** to the baseline at epoch 467.

**Why so few epochs?** The PR specifies no `--max-eval-batches`, meaning full chunked evaluation every epoch. With 16k eval surface points, each validation sample is split into ceil(N/16000) chunks evaluated sequentially. This makes each epoch ~21 minutes, vs the baseline which uses `--max-eval-batches 200` to cap eval cost. The baseline can complete many more epochs in the same time budget.

The learning curve shows the model was at the start of its second cosine cycle (T_max=30), with best metrics at cosine minima (ep15: 25.6%, ep17: 24.0%) — consistent with a model that would continue improving substantially with more training.

### Details

- **W&B run:** [`icqb9nla`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/icqb9nla)
- **Peak GPU memory:** ~2.88 GB (2.9% of 97.9 GB RTX PRO 6000)
- **Command:**
  ```bash
  cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
    --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
    --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
    --model-heads 8 --epochs 999 --batch-size 1 \
    --drivaerml-train-surface-points 16000 --drivaerml-eval-surface-points 16000 \
    --max-train-batches 394 \
    --wandb-group dm-surface-points-sweep \
    --agent alphonse --wandb-name "alphonse/dm-16k-surface-points"
  ```

### What happened

The 16k surface point config itself ran fine — no OOM, low memory usage (2.88 GB), training loss was decreasing healthily. The fundamental issue is that **full eval without `--max-eval-batches` is prohibitively expensive**, consuming most of the epoch time and limiting the run to just 17 epochs in the 360-minute window. The model needed hundreds more epochs to converge.

The very low memory (2.88 GB vs 97.9 GB available) confirms that 16k points dramatically reduces VRAM requirements — which could enable larger batch sizes or model dimensions.

### Suggested follow-ups

1. **Re-run with `--max-eval-batches 200`** to match the baseline's eval budget and allow fair comparison over many more epochs. This is the critical change needed for a meaningful comparison.
2. **Alternatively, increase timeout** significantly if full eval is truly required — the run would need ~60+ hours to reach 200+ epochs at current speed.
3. The low VRAM usage (2.88 GB) suggests 16k points could pair well with larger model configs or batch accumulation steps, if the epoch throughput issue is resolved.

**Comment by tcapelle** [comment]
ADVISOR: Same eval-overhead issue as 32k/64k — without --max-eval-batches, full chunked eval costs ~21 min/epoch, leaving only 17 epochs in the timeout (vs 467 needed for convergence). Please retry with **--max-eval-batches 200** added. Keep all other settings the same (16k surface points, champion LR/T_max/architecture). Note: the very low VRAM (2.88 GB vs 97.9 GB available) is interesting — once you have the baseline 16k result, a follow-up pairing 16k points with a larger model or batch accumulation could be valuable.

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results (Retry with --max-eval-batches 200)

### DrivAerML — 16k surface points (train + eval), max-eval-batches 200

| Metric | Baseline (50k pts, ep467) | This run (16k pts, ep93) |
|--------|--------------------------|--------------------------|
| Val `surface_rel_l2_pct` | **3.997%** | 11.672% |
| Test `surface_rel_l2_pct` | 6.244% | 13.827% |
| Best epoch | 467 | 93 |

**External targets:** AB-UPT 3.82% | Transolver-3 3.71%

### Key finding: catastrophic instability with 16k surface points

Adding `--max-eval-batches 200` fixed the epoch throughput problem — the run completed 526 epochs (vs 17 without the flag). However, the 16k surface point config proved **fundamentally unstable** with the cosine LR schedule (T_max=30):

1. **Epochs 1–93: healthy convergence.** Val improved from 64.9% → 11.7%, with clear improvements at each cosine trough:
   - Trough 1 (ep15): 23.6% → Trough 2 (ep30): 19.6% → Trough 3 (ep45): 16.9% → Trough 4 (ep60): 14.5% → Trough 6 (ep90): 12.0%

2. **Epochs 93–209: recurring catastrophic spikes.** The model suffered 5 major divergence events (epochs 128, 163, 174, 187, 200) where val spiked to 40–80%. Each spike partially recovered but prevented the model from improving past its epoch-93 best of 11.7%. The spikes occurred at LR peaks in the cosine schedule, suggesting 16k points provides insufficient gradient signal to stabilize through high-LR phases.

3. **Epoch 209+: terminal NaN divergence.** Gradient norms exploded (44T → 137T → NaN at epochs 207–209) and the model never recovered. 317 NaN epochs followed until timeout.

### Why 16k fails

The cosine LR schedule with T_max=30 cycles the learning rate up to 5e-4 every 30 epochs. With 50k surface points, each training sample provides dense enough gradient signal to keep the model stable through these LR peaks. With only 16k points, the sparser sampling creates noisier gradients that can't sustain the high-LR phases, leading to periodic catastrophic instability and eventual divergence.

### Details

- **W&B run:** [`oh89yggk`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/oh89yggk)
- **Peak GPU memory:** ~2.7 GB (2.8% of 97.9 GB RTX PRO 6000)
- **Total epochs:** 526 (209 valid + 317 NaN)
- **Command:**
  ```bash
  cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
    --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
    --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
    --model-heads 8 --epochs 999 --batch-size 1 \
    --drivaerml-train-surface-points 16000 --drivaerml-eval-surface-points 16000 \
    --max-train-batches 394 --max-eval-batches 200 \
    --wandb-group dm-surface-points-sweep \
    --agent alphonse --wandb-name "alphonse/dm-16k-surface-points-maxeval200"
  ```

### What happened

The `--max-eval-batches 200` fix allowed sufficient epochs for a fair comparison, but 16k surface points is conclusively too sparse for stable training with this LR schedule. The model achieved val 11.7% / test 13.8% before diverging — roughly 3× worse than the 50k baseline (val 4.0% / test 6.2%). The very low VRAM (2.7 GB) confirms that 16k dramatically reduces memory, but at the cost of training stability.

### Suggested follow-ups

1. **16k with lower peak LR (e.g. 2e-4 or 1e-4)** — the instability is clearly LR-peak-driven; a lower max LR may prevent the catastrophic spikes while retaining the VRAM savings.
2. **16k with longer T_max (e.g. 60 or 100)** — gentler LR cycling could avoid the sharp peaks that trigger divergence.
3. **16k with gradient clipping** — could prevent the gradient explosions that lead to NaN.
4. **32k surface points** — a middle ground that may balance stability with reduced computation.
5. **16k with larger batch size** — the low VRAM (2.7 GB) leaves room for batch_size=2+, which would average out gradient noise.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3066 because ADVISOR: 16k surface points dead — 11.672% best (3x worse), terminal instability at ep209. Sparser gradients cannot survive cosine LR cycling. Combined with 32k (7.558%) and 64k (diverges), fewer-than-50k surface points are conclusively dead for DM. Reassigning to noam-feature experiment..
